CGNFY-58: Process Quest

amenasyed opened the pull request:
## Ticket
[CGFNY-58](https://asma71612.atlassian.net/jira/software/projects/CGNFY/boards/1?selectedIssue=CGNFY-58)

## Description
- Set up an AWS S3 bucket for storing raw audio (in .wav format ) and transcription files (in .json format)
- Configured backend server using Express for uploading audio to S3 for transcription and fetching the transcription result 
- Implemented Process Quest Game UI and sent relevant fields back to Firebase for Reporting
- Generated procedural narrative prompts for demo
- Added instruction page for Process Quest with feedback implemented from stakeholders
- Linked Cognify logo on patient/physician login pages to homepage for ease of navigation

## Testing 
**STEP 0: ENSURE YOU HAVE GONE THROUGH [THE AWS SETUP INSTRUCTIONS](https://github.com/asma71612/cognitive-decline-monitoring?tab=readme-ov-file#step-9-aws-setup) AS DETAILED IN THE README**

1. Create a new test user in Cognify from the Physician side (**My Patients > Add Patient**)
2. Login using generated test ID.
3. Click on "Complete My Task"
4. You will be lead through the first part of Memory Vault (this has already been tested). Once completed, click "Next".
5. This screen should resemble **SCREEN 1**. Click "Start"
6. This screen should resemble **SCREEN 2**. Depending on your playCount index, a different prompt will appear accordingly. A timer is in the top-right corner decreasing from 15 minutes. After responding to the prompt, click "Next".
7. Complete the second part of Memory Vault and click "Done".
8. To verify reports are sent, go to Firebase and ensure the payload matches user with test ID **"ID-ybi7ioil8nq"** (this should be populated for Memory Vault _and_ Process Quest since you completed both)
9. To confirm reporting is generated, go to **My Reports** (from patient homepage). Verify that you can see both Memory Vault and Process Quest daily reports. 
10. Scroll to the bottom of Daily Reports and click "See More". Ensure that you can see metrics for both Memory Vault and Process Quest.
11. Log out of test patient
12. Log into a physician account
13. Navigate to My Patients > Test Patient > "View Reports"
14. Ensure you can see Daily Reports, Daily Reports See More and All Time Reports metrics populated properly. See **VIDEO 1** for reference.

**OTHER TESTING**
- You can decrease the seconds on the timer to ensure that recording stops and navigation to the next screen triggers when the timer is up
- Ensure that Cognify logo on patient/physician login pages links back to homepage
- Navigate to S3 bucket and ensure the audio (wav) and transcript (json) files uploaded for the session follow the path "processQuest_{user-id}_{date}_session{session-number}"

**SCREEN 1: INSTRUCTIONS**
<img width="1438" alt="screen1" src="https://github.com/user-attachments/assets/d3ca7882-4a34-447d-84c3-1714a2664efc" />

**SCREEN 2: GAME UI**
<img width="1438" alt="screen2" src="https://github.com/user-attachments/assets/91396eae-6a16-4346-90bb-ef318f2990cd" />

**VIDEO 1: METRICS FROM PATIENT AND PHYSICIAN SIDES** _(note as confirmed with teammates that some metrics were changed/omitted and some of the reporting side needs modification)_
https://drive.google.com/file/d/1Er-lWaSvoNDyE9HzKkmqpAxZqiK7_sCK/view?usp=sharing

## Checklist
- [ ] I have performed a self-review of my code.
- [ ] I have added tests that prove my fix is effective or my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added documentation where necessary.

a-ansari02 commented:
I have a few comments in general before I go in and review your PR officially:

**`README COMMENT 1`**
the command you tell us to run in the readme (`echo "test file" > test.txt aws s3 cp test.txt s3://cognify-capstone/`) are actually 2 separate commands and if we run them together, they create a file locally instead and do not upload anything to the bucket.
```bash
# First create the test file
echo "test file" > test.txt

# Then upload it to S3
aws s3 cp test.txt s3://cognify-capstone/
```
**but there's a problem here because this is creating the file locally and crowding our local repo - can we find another solution to do this?**

**`README COMMENT 2`**
I think the instructions could be fixed up a little like navigating the permissions and policies because when I was following the instructions, it was really hard to figure out the steps and some of the things it was saying to do, was not really true to what is displayed in the console. 

**`README COMMENT 3`**
We should add a warning at the top of the step 9 instructions that we need to make sure to have `server.cjs` file and all latest changes locally and then follow these steps. Good to give a quick disclaimer.

**`README COMMENT 4`**
Add these lines in because we are adding new python dependencies and I had trouble with these imports:
```bash
pip install pandas
pip install openpyxl
```

**`README COMMENT 5`**
The way it says to setup the policies are a bit contradictory because it says to give s3 and transcribe full access but then we are basically overwriting the s3 access by editing the json of the policy in the next step.

**`MEMORY VAULT`**
We need to revisit finding a way to add a buffer to this because I cant understand this audio:

https://github.com/user-attachments/assets/2b26102b-d013-4daf-b09a-2932de288301

**`PROCESS QUEST`**
~~Biggest concern of them all: The transcripts are not generating, therefore the report is not generating. Now I don't know what the issue is but I'd advise you to double check that you provided ALL the setup instructions (any missing transcribe policies that need to be attached? anything in the code that is not saving the transcripts properly? though i suspect its not a code issue)~~

^ Did some debugging and figured out that the issue is in the README instructions. It essentially tells us to override the `Transcribe` settings which should not be happening. We need a revisit of these instructions to ensure we can all set it up correctly.

**`PERMISSIONS AND ACCESS`**
You are the owner of the bucket and have the ability to edit and play around with permissions and since this is a part of your PR, you should make the change to the S3 bucket policy as well and make sure only specific ARNs have access to them. No one else has permissions to edit those for you.

The acceptance criteria (A/C) for this is:
- Log in with an account that has **NOT** followed the setup instructions and make sure the bucket is **NOT** visible to them. 
- Log in with an account that has followed the setup instructions and make sure the bucket is visible to them.

a-ansari02 commented:
Can you also add a check `"Processing in progress. Check back in X minutes."` (something along those lines) to make sure it is not empty if the transcription takes long and I quickly click the reports button?

asma71612 commented:
Also update https://github.com/asma71612/cognitive-decline-monitoring?tab=readme-ov-file#prerequisites and https://github.com/asma71612/cognitive-decline-monitoring?tab=readme-ov-file#current-versions sections in the README

amenasyed commented:
@a-ansari02 I addressed all your pre-PR comments with a few notes:
- Can you confirm which user policies were giving you issues? And how did you solve them? This is helpful to know so I can update the README accordingly
- I dont have any reference to `openpyxl` locally nor have I introduced it in any of my changes
- I have uploaded new audio for a few words that were hard to distinguish in Memory Vault (egg, boat)
- Confirmed with @slollino that only specified ARNs have access to the S3 bucket

a-ansari02 commented:
@amenasyed 

1. I saw you added the ARNs line in the README so technically we don't really need to mess with the policies too much.
2. Gotcha, not sure why I was getting that error but thanks for verifying.
3. Perfect, they sound clearer now.
4. 👍 

> Can you also add a check "Processing in progress. Check back in X minutes." (something along those lines) to make sure it is not empty if the transcription takes long and I quickly click the reports button?

^ Regarding this comment, for some clarification, we will likely have cases where users will spend that entire 10-15 minutes talking in which case the processing will take much longer than we've been testing (at least I haven't tested for longer than 1 minute). I was going to do some pressure testing here but something seems wrong with transcriptions so that will need to be fixed. But the point is that we want to see if the data has finished uploading to s3 and firebase and if it has not, then we display a message on the reporting side saying `"Check back... etc."`

Before you re-request review on any iterations of your PR, it'll be good if you do end to end testing yourself to make sure nothing is broken and the pressure testing works and everything is in order. Then if I encounter any issues, I'll know it's my own setup vs spending a long time debugging not knowing if the issue is on my end or your end.

amenasyed commented on react_frontend/src/pages/games/processQuest/ProcessQuest.js:
```diff
@@ -0,0 +1,336 @@
+    startRecording();
+
+    return () => stopRecording();
+  }, [userId]);
```
I tested this and since we're updating the aws job name with a unique date identifier, a new request is sent every time Date.now() changes. So this is not a needed dependency.
